[scrollbar-styling] Scrollbar-gutter should affect scroll container's intrinsic size

nmoucht opened the pull request:
#### 17701939e9dd2895c1f80708f9f6db0241a69b07
<pre>
[scrollbar-styling] Scrollbar-gutter should affect scroll container&apos;s intrinsic size
<a href="https://bugs.webkit.org/show_bug.cgi?id=277047">https://bugs.webkit.org/show_bug.cgi?id=277047</a>
<a href="https://rdar.apple.com/132509923">rdar://132509923</a>

Reviewed by Simon Fraser.

Scrollbar gutter should affect the intrinsic size of its container based on the chart in the spec:
<a href="https://www.w3.org/TR/css-overflow-3/#scrollbar-gutter-property.">https://www.w3.org/TR/css-overflow-3/#scrollbar-gutter-property.</a> Also remove asserts that are no
longer necessary because there are cases where we want to show scrollbar-gutter even when we do not
display the overlay scrollbars.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeIntrinsicLogicalWidths const):
(WebCore::RenderBlock::computePreferredWidthsForExcludedChildren const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::verticalScrollbarWidth const):
(WebCore::shouldShowScrollbarGutter):
(WebCore::RenderBox::intrinsicScrollbarLogicalWidth const):
(WebCore::RenderBox::repaintDuringLayoutIfMoved):
* Source/WebCore/rendering/RenderBoxInlines.h:
(WebCore::RenderBox::contentBoxRect const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const):

Canonical link: <a href="https://commits.webkit.org/281527@main">https://commits.webkit.org/281527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/698d49dce3dc20ef975c807c7d9f25fb7c459031

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60170 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12722 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64088 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10700 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62300 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48739 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7468 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62201 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36841 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52122 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29581 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33549 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9620 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55462 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9634 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65820 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4100 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9516 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56094 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56249 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3413 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9027 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35331 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36413 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37501 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -954,20 +954,21 @@ int RenderBox::horizontalScrollbarHeight() const
     return includeHorizontalScrollbarSize() ? scrollableArea->horizontalScrollbarHeight(IgnoreOverlayScrollbarSize, isHorizontalWritingMode()) : 0;
 }
 
+static bool shouldShowScrollbarGutter(ScrollbarGutter gutter, bool hasVisibleOverflow, Overflow overflow)
```
This could be a lambda function.

It's a bit odd to say "show" for something that's just a gap.

nmoucht replied:
Changed to lambda. Changed the name to `shouldIncludeScrollbarGutter`.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -954,20 +954,21 @@ int RenderBox::horizontalScrollbarHeight() const
+    return overflow == Overflow::Scroll || (overflow == Overflow::Auto && (!gutter.isAuto || hasVisibleOverflow)) || (overflow == Overflow::Hidden && !gutter.isAuto);
+}
+
 int RenderBox::intrinsicScrollbarLogicalWidth() const
```
It's not quite right to sometimes return ` verticalScrollbarWidth() * 2` from a function that purportedly returns the scrollbar width; it would be easy to write code that assumes that the return value is actually for the scrollbar. Either rename this or put the gutter code somewhere else.

nmoucht replied:
Changed the name to `intrinsicScrollbarLogicalWidthIncludingGutter`.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -954,20 +954,21 @@ int RenderBox::horizontalScrollbarHeight() const
-        ASSERT(layer() && layer()->scrollableArea() && layer()->scrollableArea()->hasVerticalScrollbar());
-        return verticalScrollbarWidth();
-    }
+    if (isHorizontalWritingMode() && ( ( style().overflowY() == Overflow::Scroll || shouldShowScrollbarGutter(style().scrollbarGutter(), hasScrollableOverflowY(), style().overflowY()) ) && !canUseOverlayScrollbars()))
```
Extra spaces

nmoucht replied:
Fixed.

nmoucht committed:
[scrollbar-styling] Scrollbar-gutter should affect scroll container's intrinsic size
https://bugs.webkit.org/show_bug.cgi?id=277047
rdar://132509923

Reviewed by Simon Fraser.

Scrollbar gutter should affect the intrinsic size of its container based on the chart in the spec:
https://www.w3.org/TR/css-overflow-3/#scrollbar-gutter-property. Also remove asserts that are no
longer necessary because there are cases where we want to show scrollbar-gutter even when we do not
display the overlay scrollbars.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeIntrinsicLogicalWidths const):
(WebCore::RenderBlock::computePreferredWidthsForExcludedChildren const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::verticalScrollbarWidth const):
(WebCore::shouldShowScrollbarGutter):
(WebCore::RenderBox::intrinsicScrollbarLogicalWidth const):
(WebCore::RenderBox::repaintDuringLayoutIfMoved):
* Source/WebCore/rendering/RenderBoxInlines.h:
(WebCore::RenderBox::contentBoxRect const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const):

Canonical link: https://commits.webkit.org/281527@main
